src/common/w25x.cpp

```diff
@@ -0,0 +1,474 @@
+    uint32_t addr_align = addr % PAGE_SIZE;
+    if (addr_align != 0) {
+        uint32_t cnt_align = PAGE_SIZE - addr_align;
+        if (cnt_align >= cnt) {
+            program_page(addr, data, cnt, high_priority);
+            return;
+        }
+        program_page(addr, data, cnt_align, high_priority);
+        addr += cnt_align;
+        data += cnt_align;
+        cnt -= cnt_align;
+    }
+
+    // Write all full pages
+    while (cnt >= PAGE_SIZE) {
+        program_page(addr, data, PAGE_SIZE, high_priority);
+        addr += PAGE_SIZE;
+        data += PAGE_SIZE;
+        cnt -= PAGE_SIZE;
+    }
+
+    // Write the remaining data
+    if (cnt > 0) {
+        program_page(addr, data, cnt, high_priority);
+    }
+}
+
+static constexpr uint32_t tSUS_ns = 20000;
+
+static void suspend_erase() {
+    w25x_select();
+    w25x_send_byte(CMD_ERASE_PROGRAM_SUSPEND);
+    w25x_deselect();
+    // W25Q guarantees to be available in tSUS
+    // alternatively busy status can be polled
+    DELAY_NS_PRECISE(tSUS_ns);
+}
+
+static void resume_erase() {
+    w25x_select();
+    w25x_send_byte(CMD_ERASE_PROGRAM_RESUME);
+    w25x_deselect();
+    // Assure that suspend is not called earlier than in tSUS
+    // after resume
+    DELAY_NS_PRECISE(tSUS_ns);
+}
+
+void w25x_program(uint32_t addr, const uint8_t *data, uint32_t cnt) {
+    const bool high_priority = (addr >= w25x_pp_start_address) && (addr < w25x_fs_start_address);
+    const bool wait = !high_priority;
+
+    OptionalMutex eraseMutex(high_priority ? erase_mutex : NULL, wait);
+    {
+        OptionalMutex communicationMutex(high_priority ? communication_mutex : NULL);
+        if (is_erasing) {
+            suspend_erase();
+        }
+        split_page_program(addr, data, cnt, high_priority);
+
+        if (is_erasing) {
+            resume_erase();
+        }
+    }
+}
+
+static void w25x_erase(uint8_t cmd, uint32_t addr) {
+    OptionalMutex eraseMutex(erase_mutex);
+    {
+        OptionalMutex communicationMutex(communication_mutex);
+        is_erasing = true;
+        w25x_enable_wr();
+        w25x_select();
+        CmdWithAddress cmdWithAddress = cmd_with_address(cmd, addr);
+        w25x_send(cmdWithAddress.buffer, sizeof(cmdWithAddress.buffer));
+        w25x_deselect();
+    }
+    if (!w25x_wait_erase())
+        w25x_set_error(HAL_TIMEOUT);
+}
+
+void w25x_sector_erase(uint32_t addr) {
+    w25x_erase(CMD_SECTOR_ERASE, addr);
+}
+
+void w25x_block32_erase(uint32_t addr) {
+    w25x_erase(CMD_BLOCK32_ERASE, addr);
+}
+
+void w25x_block64_erase(uint32_t addr) {
+    w25x_erase(CMD_BLOCK64_ERASE, addr);
+}
+
+void w25x_chip_erase(void) {
+    OptionalMutex eraseMutex(erase_mutex);
+    {
+        OptionalMutex communicationMutex(communication_mutex);
+        is_erasing = true;
+        w25x_enable_wr();
+        w25x_select();
+        w25x_send_byte(CMD_CHIP_ERASE); // send command 0xc7
+        w25x_deselect();
+        if (!w25x_wait_busy())
```

Comment:
How does the is_erasing flag get set to `false` here?

Reply:
Good catch. This flag needs to be removed. Thanks to my stupidity I didn't noticed for the first time that it doesn't have any sense to postpone chip erase operation. So initially I wrapped it the same way as all block erase operations. While reviewing the datasheet again I noticed this is (logically) not supported. And I forget to remove this flag while rewriting that to totally blocking operation.

Reply:
Removed.

Reply:
And renamed to `is_erasing_block` to make it more clear.